dr: introduce the dr command

nirs opened the pull request:
The "odf dr" command exposes selected ramenctl[1] commands that are suitable to use with ODF.

The command support the following sub-commands:

- init: Create configuration file for dr commands
- test: Test disaster recovery flow with a tiny application

The command is using the ramenctl/cmd.RootCmd with modified name and description, and selected sub commands.

We annotate the command to let ramenctl know that it is running as "odf dr". The name is used in the configuration file.

Running "odf dr" commands is identical to running "ramenctl" command. The configuration file is used to adapt the command to different Kubernetes distribution (distro), storage providers (pvcSpecs), OCM configuration (clsuterSet) and DR configuration (drPolicy).

Issues:

- The init command defaults should be improved when running as "odf dr" to minimize manual configuration.
- ramenctl/pkg/build should contain the version and git commit hash for dr commands reports. This information seems to be missing in odf cli.
- The globals flags inherited from the RootCmd are ignored by the dr command. There seems to be no good way to fix this in ramenctl.

[1] https://github.com/RamenDR/ramenctl

subhamkrai reviewed (changes requested):
@nirs  Thanks for the PR but could you add this first in upstream https://github.com/rook/kubectl-rook-ceph and then just import the pkgs here similar to the other commands.

nirs commented:
> @nirs Thanks for the PR but could you add this first in upstream https://github.com/rook/kubectl-rook-ceph and then just import the pkgs here similar to the other commands.

No, this command does not make sense for kubectl-rook-ceph. It is only for the odf tool.

nirs commented on cmd/odf/dr/dr.go:
```diff
@@ -0,0 +1,29 @@
+
+		build.Version = "v1.2.3"
+		build.Commit = "eb92ed81e2715d286bfd8ce173c76d4ecda9e2b4"
+	*/
```
In ramenctl we add version and commit hash when building the tool:
https://github.com/RamenDR/ramenctl/blob/main/Makefile

So in the report we show:

```yaml
ramenctl:
  commit: b5e7ffde13922f0fe0695f404026cd27bb33c34b
  version: v0.1.0-19-gb5e7ffd
```

nirs replied:
This will be optional in next release:
https://github.com/RamenDR/ramenctl/pull/63

If we add build info we can expose it in the report.

nirs replied:
The "ramenctl" part is now "build":

```yaml
build:
  version: v0.3.0
  commit: 78e90bed8fa6d2568ea4d7184b4f0de1870349a5
```

And it is omitted if we don't have build info.

nirs commented on cmd/odf/dr/dr.go:
```diff
@@ -0,0 +1,29 @@
+
+func init() {
+	// Modify ramenctl RootCmd for odf-cli
+	DrCmd.Use = "dr"
```
I think "odf dr" is the best name for this command.

`odf get dr-health` and `odf get dr-prereq` are indeed little bit confusing, mainly because they are not part of "odf dr". However they not really about dr but about storage replication health, which is not in scope for ramenctl.

We can consider renaming them, maybe adding an alias so the old name keeps working for now.

nirs replied:
So I'll keep "dr" for the name. We start to work on the validate command next week, and I'll check how can integrate getting storage health there.

I don't think we want a simple alias running current `get dr-health` as `dr health`, since the current command works with single cluster (using --context, --kubeconfig, --namespaces, ...), and all dr commands works with a config file with all clusters. So `dr health` should check the status for both clusters, report the result in the command report, and display it in a consistent way with other validations.

It makes sense to keep the current commands for backward compatibility and for case when. you want to do specific check on single cluster.

nirs commented on docs/dr.md:
````diff
@@ -0,0 +1,152 @@
+The command stores `test-run.yaml` and `test-run.log` in the specified directory:
+
+```bash
+% tree odf-dr-test/
````
`tree odf-dr-test/` is the command, we want to show it different from the output. I will change it to the common style:

```bash
$ tree odf-dr-test/
...
```

nirs commented on go.mod:
```diff
@@ -6,6 +6,7 @@ toolchain go1.23.4
 
 require (
 	github.com/pkg/errors v0.9.1
+	github.com/ramendr/ramenctl v0.2.0
```
We will have new release tomorrow, will update in the next push.

nirs commented on docs/dr.md:
```diff
@@ -0,0 +1,152 @@
+    kubeconfig: /Users/nir/.config/drenv/rdr/kubeconfigs/dr1
+  c2:
+    name: dr2
+    kubeconfig: /Users/nir/.config/drenv/rdr/kubeconfigs/dr2
```
This is modified config for ramen testing env, not the default config. Need to update to show the defaults.

nirs commented on docs/dr.md:
````diff
@@ -0,0 +1,152 @@
+The test command tests complete disaster recovery flow with a tiny application.
+
+```bash
+% ramenctl test run -o odf-dr-test
````
Use the `$ ramenctl test run -o odf-dr-test`

nirs replied:
Yes I pasted the wrong example from another run with more recent output and forgot to update the command.

nirs commented:
Changes in latest version:
- Updated to latest ramenctl release (v0.3.1)
- Remove the TODO for the build version, build info is omitted if not available.
  - If we add build info during the build, we can update ramenctl/pkg/build to display the build into in the report
- Fixed `ramenctl` in the example to `odf dr`
- Use the convention of `$` to identify a command instead of `%` (macOS default prompt)
- Improve comments style in the dr package
- Update the examples in `docs/dr.md` to latest release output
- Show example `test-run.yaml`
- Show example report after `test clean` command
- Add a note to run test clean after test run

Tested with OpenShift cluster using `odf dr` (the examples uses output from this test).

Example test report:
[odf-dr-test.tar.gz](https://github.com/user-attachments/files/19523102/odf-dr-test.tar.gz)

nirs commented:
Change is last version:
- Update README.md: describe the dr sub-comamnds and link to docs/dr.md.
